engine/class_modules/priest/sc_priest.cpp

```diff
@@ -353,6 +353,20 @@ struct fae_guardians_t final : public priest_spell_t
   }
 };
 
+struct wrathful_faerie_t final : public priest_spell_t
+{
+  double insanity_gain;
+
+  wrathful_faerie_t( priest_t& p )
+    : priest_spell_t( "wrathful_faerie", p, p.find_spell( 342132 ) ),
+      insanity_gain( p.find_spell( 327703 )->effectN( 2 ).resource( RESOURCE_INSANITY ) )
+  {
+    energize_type     = action_energize::ON_HIT;
+    energize_resource = RESOURCE_INSANITY;
+    energize_amount   = insanity_gain;
```

Comment:
for some reason this is not respecting the ICD of the action, i suspect its because im calling the action with `active_spells.wrathful_faerie->execute();` which may be circumventing the ICD? but im not sure.

```
Name             : Wrathful Faerie (desc=Night Fae) (id=342132) [Spell Family (6)] 
Class            : Priest
School           : Nature
Spell Type       : None
Range            : 100 yards
Max Targets      : 1
Velocity         : 60 yards/sec
Duration         : 20 seconds
Internal Cooldown: 1.5 seconds
Proc Chance      : 100%
Proc Flags       : ........ .....x.. .x...... ........ 
                 : Generic Hostile Spell Taken, Magic Hostile Spell Taken
```

You can see that even though spelldata has a 1.5s ICD the sim doesn't respect it:

```
9.958 Base gains 3.00 (3.00) insanity from wrathful_faerie (70.00/100.00)
10.924 Base gains 3.00 (3.00) insanity from wrathful_faerie (77.00/100.00)
```

Trying to set it manually also gives the same result:

```
internal_cooldown->duration = timespan_t::from_seconds( 1.5 );
```

Reply:
@scamille any ideas here?

Reply:
Afaik background actions don't get any cooldown restrictions.
you could provide custom cooldown activation by overriding update_ready.
Or maybe something completely different than a normal action, with hand-crafted cooldown and generate insanity code.

Reply:
I took a bit of a mix from suggestions from you and Ansh here and came up with this:

Set the cd to the ICD:
```
cooldown->duration = data().internal_cooldown();
```

Create a custom trigger function that we can start/check the CD:
```
void trigger()
  {
    if ( priest().cooldowns.wrathful_faerie->is_ready() )
    {
      execute();
      priest().cooldowns.wrathful_faerie->start();
    }
  }
```

Looks good: (not exactly 1.5s because of the GCD on casting spells)
```
5.131 Base gains 3.00 (3.00) insanity from wrathful_faerie (29.00/100.00)
7.065 Base gains 3.00 (3.00) insanity from wrathful_faerie (40.00/100.00)
```